### PR TITLE
applications: nrf_desktop: Add genric axes support

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:DESK-879

### CODE REVIEW

**Comment by pdunaj** [comment]
Tested with mouse - connected with BLE and USB. Works fine. In release mode speed is not changed.

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 1]
```diff
@@ -989,12 +999,15 @@ static void update_key(const struct hid_keymap *map, s16_t value)
...
-	__ASSERT_NO_MSG(pos < INPUT_REPORT_COUNT);
+	struct report_data *rd = get_report_data(report_id);
+
+	bool connected = false;
 
-	struct report_data *rd = &state.report_data[pos];
-	bool connected = state.selected &&
-		(state.selected->state[pos].state != STATE_DISCONNECTED);
+	if (state.selected) {
+		struct report_state *rs = get_report_state(state.selected, report_id);
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state.c]: `__ASSERT_NO_MSG(rs != NULL)`

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 2]
```diff
@@ -478,10 +499,9 @@ static void disconnect_subscriber(const void *subscriber_id)
 				/* Clear state if notification was not
 				 * disabled before disconnection.
 				 */
-				struct report_data *rd = &state.report_data[pos];
+				struct report_data *rd = get_report_data(r_id);
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state.c]: This should be added in previous commit

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 3]
```diff
@@ -1046,6 +1044,7 @@ static void init(void)
 	if (IS_ENABLED(CONFIG_DESKTOP_HID_REPORT_MOUSE_SUPPORT)) {
 		report_index[REPORT_ID_MOUSE] = id;
 		state.report_data[id].items.item_count_max = MOUSE_REPORT_BUTTON_COUNT_MAX;
+		state.report_data[id].axes.axis_count = 3; /* X, Y, Wheel */
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state.c]: Is this used anywhere?

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 4]
```diff
@@ -696,7 +716,14 @@ static void send_report_mouse(u8_t report_id, struct report_data *rd)
 
 	EVENT_SUBMIT(event);
 
-	rd->items.update_needed = false;
+	if ((rd->axes.axis[0] != 0) ||
+	    (rd->axes.axis[1] != 0) ||
+	    (rd->axes.axis[2]  < -1) || (rd->axes.axis[2]  > 1)) {
+		/* If there is some axis data to send, request report update. */
+		rd->items.update_needed = true;
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state.c]: Maybe we can simplify it to one `update_needed` per report data?

Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_state.c]: Currently we do not use the update_needed from `rd->axes`

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 5]
```diff
@@ -83,10 +86,17 @@ struct eventq {
 	size_t len;
 };
 
-/**@brief Report state. */
+/**@brief Axis data. */
+struct axis_data {
+	s16_t axis[AXIS_COUNT]; /**< Array of axes. */
+	u8_t axis_count; /**< Number of axes in thi set. */
+	bool update_needed; /**< Set was updated, report must be generated. */
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: I prefer set as not all elements of this array are valid. Valid items are also present at the back of it. So for me it looks as a data set not as a generic array.

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 2 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: ah - true

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 3 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: not at the moment

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 4 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: will do. thanks, overlooked.

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 1 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: we cannot move the assert there as NULL is a valid return used in one case

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_state.c]: ok

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 1 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_state.c]: Ok, let's leave it as it is.

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 3 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_state.c]: So maybe we can remove it to save memory? We know how big the array is and we may make checks based on size of the array.

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 6]
```diff
@@ -83,10 +86,17 @@ struct eventq {
 	size_t len;
 };
 
-/**@brief Report state. */
+/**@brief Axis data. */
+struct axis_data {
+	s16_t axis[AXIS_COUNT]; /**< Array of axes. */
+	u8_t axis_count; /**< Number of axes in thi set. */
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: done

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 4 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: done

File: applications/nrf_desktop/src/modules/hid_state.c (hunk 2 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: Fixed

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 7]
```diff
@@ -645,17 +665,17 @@ static void send_report_mouse(u8_t report_id, struct report_data *rd)
...
 	/* X/Y axis */
-	s16_t dx = MAX(MIN(state.last_dx, MOUSE_REPORT_XY_MAX),
+	s16_t dx = MAX(MIN(rd->axes.axis[0], MOUSE_REPORT_XY_MAX),
 		       MOUSE_REPORT_XY_MIN);
-	s16_t dy = MAX(MIN(-state.last_dy, MOUSE_REPORT_XY_MAX),
+	s16_t dy = MAX(MIN(-rd->axes.axis[1], MOUSE_REPORT_XY_MAX),
 		       MOUSE_REPORT_XY_MIN);
-	state.last_dx -= dx;
-	state.last_dy += dy;
+	rd->axes.axis[0] -= dx;
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: Done

File: applications/nrf_desktop/src/modules/hid_state.c [hunk 8]
```diff
@@ -386,7 +392,22 @@ static void clear_items(struct items *items)
 {
 	memset(items->item, 0, sizeof(items->item));
 	items->item_count = 0;
-	items->update_needed = false;
+}
+
+static void clear_axes(struct axis_data *axes)
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state.c]: You clear axis_count here too, is it intended?

Comment (pdunaj) [reply applications/nrf_desktop/src/modules/hid_state.c]: no, but let's fix it in the next PR. it is not used yet

**Pull request merged**